### ISSUE TITLE
cleanfigure spinoff

### ISSUE THREAD

**Issue by okomarov** [matlab2tikz/matlab2tikz#678]
Do you mind if I create a cleanfigure spinoff (as a separate repo) and post it on the FEX?

I use cleanfigure a lot outside m2t especially for the line simplification stuff and I think that with the recent problems with HG2 performance that it might be a good contribution on its own.

**Comment by egeerardyn** [comment]
The only concern I have with that is that it would make it easier for the version in `matlab2tikz` and the separate entry to diverge. So that means bugs reported in different places, different code at both places, ...

Could you elaborate your plans? Probably we can do something similar within our community or actually even spin out `cleanfigure` into a separate repository (in the `matlab2tikz` project). I think we should be careful about splitting since this risks duplicating some effort among our limited amount of developers, for -- what I think -- is a slim benefit.

**Comment by okomarov** [comment]
I meant something simpler than that :). I want to use the `cleanfigure` codebase and make it a tool independent of m2t. Nothing will happen to the `m2t` repo, nor I am suggesting any modularization. 

Obviously, all license requirements will be satisfied and original authors mentioned.

**Comment by egeerardyn** [comment]
Can you elaborate your plans in more detail, because it's not clear to me what you actually want to do. License requirements are quite limited since its a BSD license :smile:

**Comment by okomarov** [comment]
Fork `m2t`, delete everything but `cleanfigure`, make it runnable on its own, release it on fex.

**Comment by egeerardyn** [comment]
Really fork `m2t` as in a separate GitHub repo where you intend to have its own diverging development?

Or practically just a "dumb" release (i.e. no development of its own) basically to improve visibility of the tool on FEX?

I don't like the former since that brings a slew of problems as that makes integrating changes from one side into the other quite an ordeal and it will probably halve the effort on either side. I really prefer if we can keep a single authoritative repository for `cleanfigure`, instead of two competing ones. I can also only cause confusion when development diverges (I don't see a reason to diverge the development, but maybe you have a different vision?).

I do see some ways that we can have a separate FEX entry for `cleanfigure` without getting two competing repositories: e.g. split out `cleanfigure` into a standalone repo and importing it as a dependency in `m2t` or just make some release scripts for `cleanfigure` as standalone.

But I think it is in everybody's interest to keep a single authoritative repository for `cleanfigure` instead of two competing variants. So I hope we can continue this as a coordinated team effort or at least jointly decide upon what the best course of action is.

**Comment by PeterPablo** [comment]
I share the concerns of @egeerardyn. As far as I understood your usecase, you would like to make `cleanfigure` work standalone. So why not adapting both `m2t` and `cleanfigure` in a way that `cleanfigure` is extended by your standalone feature(s) and make `m2t` aware of the changes in a way that the current call syntax either remains valid or is marked deprecated and `m2t` is adapted to the new interface?! 

Apart from that `cleanfigure` could be split into a separate repository under the GitHub user `matlab2tikz` to emphasize both it's context and usecase.

@okomarov, would that work?

**Comment by egeerardyn** [comment]
I agree with @PeterPablo and I think there are certainly some things we can do with `cleanfigure` to market that as a standalone tool, but I think we should act prudently to avoid creating two incompatible codebases.

But perhaps it's more productive to discuss this in a faster medium such as HipChat, such that we can bounce ideas back-and-forth with a lot more ease?

**Comment by okomarov** [comment]
I think there is opportunity for cleanfigure to be a successful standalone submission but there is limited time to gain visibility until TMW fixes the problem with graphic performance in HG2. 

The performance problems were raised in long posts from Answers (www.mathworks.com/matlabcentral/newsreader/view_thread/337755) and the Newsgroup and got particular attention on the blog (and semi-ufficial comunications by TMW team):  http://blogs.mathworks.com/graphics/2015/06/09/object-creation-performance/

Also, to highlight the priority of graphic performance and its relation to quantity of plotted data, the Pick-Of-The-Week recently selected a down-sampling procedure: http://blogs.mathworks.com/pick/2015/02/20/downsampling-polygons-part-2/

Additionally, hashes are poorly suited for evaluating downsampling performance, since a minimal change in the algo flags it as false positive. I mean that I really need to see what the algo is doing, so the additional layer of hash checking is superfluous. 

Git has submodules, that integrate well among different repos.

**Comment by egeerardyn** [comment]
@okomarov: That's indeed a good point. However, I disagree on the timeframe: I think it will take quite some time to speed up the HG2 framework from TMW's side. But I will try to confirm the results you mention across R2014a through R2015b(PR) to check my prejudice.

In any case, I wholeheartedly agree with you that there is an important place for `cleanfigure` to exist as an independent product and we should indeed proceed to factor it out of `m2t`. I also agree with you that hashes are a terrible metric to test that part (even for `m2t` it's seriously sub-optimal, but the best we currently have). I think there will always remain a good use-case for `cleanfigure`: a lot of data produces large figures (e.g. megabytes), which causes problems in many places: some publishers impose a maximum file size and some PDF viewers, presentation software, ... are unacceptably slow when you have such a figure in your document. Reducing the data size is hence not a LaTeX/pgfplots-only concern and one that is likely to stay relevant.

I have been thinking about a better way to evaluate the performance of the simplification: we should oversample the reduced dataset back to the original size (between the points, linear interpolation will do) and then we can use typical metrics (e.g. 2-norm, infinity-norm, ...). Probably we can include the number of (reduced) data points to compare different algorithms. But in no way is testing of `cleanfigure` insurmountable. Regardless of factoring it out of `m2t`, we should work on properly testing `cleanfigure`.

`git submodule` is a good candidate (an alternative is `git subtree`) to implement all of this. I do have some experience with `git submodule`, so I know it can be somewhat cumbersome at times (but I have no idea whether subtrees are any better). What we do have to look at, as well, is how this interacts with our updater (i.e. does github serve the submodule with the ZIP or not).

If you really are convinced that timing is of the essence, an intermediate step could be to first issue a dumb release. I.e. just a ZIP file with a working `cleanfigure` and its dependencies, a small README and that's it.

**Comment by egeerardyn** [comment]
According to http://stackoverflow.com/questions/12936014/include-a-github-submodule-automatically-in-download and https://github.com/adobe/brackets/issues/9068 and some other sources (most dating back to 2012, but the latest is 2014), GitHub does not serve submodules in its ZIP/TAR.GZ files.

This mean that we would need to make our users aware of this change. I.e. the soonest we could move `cleanfigure` into a separate module, would be the 2.0 release. I don't think it should be a deal breaker, but it will certainly increase the complexity of the project if we want to keep auto-updating functionality (and otherwise, we might want to see whether we can use git to do the updating for our users too: since R2014b, there is some git support in MATLAB (but I don't think its binaries are installed by default)).
